src/context.rs

```diff
@@ -314,13 +311,21 @@ impl Context {
             // Proxy everything by default
             None => true,
             Some(ref a) => {
-                // do the reverse lookup in our local cache
-                let mut reverse_lookup_cache = self.reverse_lookup_cache.lock();
-                // if a qname is found
-                if let Some(forward) = reverse_lookup_cache.get(ip) {
-                    !*forward
+                // first check the IP list
+                if a.check_ip_in_proxy_list(ip) {
+                    true
```

Comment:
This would lead to inconsistent behavior as `check_ip_in_proxy_list` does not distinguish whether some rule is matched or default rule is used.

Reply:
For IP rules, I think this ensures `proxy_list` in `bypass_all` and `bypass_list` in `proxy_all` have the highest priority. Otherwise we would see default domain name rules overwrite IP rules in those modes.

We can think about this, as there maybe better solutions.

Reply:
I think cache should take precedence as cache directly corresponding to a hostname match.

Reply:
IMO, any hostname resolved by DNS relay should be cached, otherwise we cannot apply hostname rules to all of them.

Reply:
Yeah, it maybe doable.

Reply:
https://github.com/shadowsocks/shadowsocks-rust/commit/ee1fd01698ca807e2aa39431e0a5b7b37058dac8